File: Data_Structures/binary_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None
        

class Tree:
    def __init__(self):
        self.root = None
        
    def _find(self, node, parent, value):
        if node is None:
            return node, parent, False
            
        if value == node.data:
            return node, parent, True
            
        if value < node.data:
            if node.left:
                return self._find(node.left, node, value)
            return node, parent, False  # Возвращаем текущий узел как потенциального родителя
                
        if value > node.data:
            if node.right:
                return self._find(node.right, node, value)
            return node, parent, False  # Возвращаем текущий узел как потенциального родителя
                
        return node, parent, False
    
    def append(self, obj):
        if self.root is None:
            self.root = obj
            return obj
            
        s, p, fl_find = self._find(self.root, None, obj.data)
        
        if not fl_find and s:
            if obj.data < s.data:
                s.left = obj
            else:
                s.right = obj
        
        return obj
    
    def show_tree(self, node):
        if node is None:
            return
        
        self.show_tree(node.left)
        print(node.data)
        self.show_tree(node.right)
    
    def _del_leaf(self, s, p):
        if p is None:  # Если удаляем единственный корень-лист
            self.root = None
        elif p.left == s:
            p.left = None
        else:
            p.right = None
    
    def _del_one_child(self, s, p):
        # Находим, какой именно потомок есть у узла s
        child = s.left if s.left is not None else s.right
        
        if p is None:  # Если удаляем корень, у которого только один потомок
            self.root = child
        elif p.left == s:
            p.left = child
        else:
            p.right = child
    
    def _find_min(self, node, parent):
        if node.left:
            return self._find_min(node.left, node) # Исправлено имя метода
        return node, parent
       
    def del_node(self, key):
        s, p, fl_find = self._find(self.root, None, key) # Исправлено obj.data на key
        
        if not fl_find:
            return None
            
        # Случай 1: Узел — лист
        if s.left is None and s.right is None:
            self._del_leaf(s, p)
            
        # Случай 2: У узла только один потомок
        elif s.left is None or s.right is None:
            self._del_one_child(s, p)
            
        # Случай 3: У узла два потомка
        else:
            sr, pr = self._find_min(s.right, s)
            s.data = sr.data
            # Правый минимальный узел (sr) гарантированно имеет не более 1 потомка (правого)
            if sr.left is None and sr.right is None:
                self._del_leaf(sr, pr)
            else:
                self._del_one_child(sr, pr)
    
    def show_level_tree(self, node):
        if node is None:
            return
        
        v = [node]
        while v:
            vn = []
            for x in v:
                print(x.data, end=' ')
                if x.left:
                    vn.append(x.left)
                if x.right:
                    vn.append(x.right)
            print()
            v = vn
            
# Проверка работы
v = [10, 5, 7, 16, 13, 2, 20]
t = Tree()
for x in v:
    t.append(Node(x))

print("Дерево до удаления:")
t.show_tree(t.root)
print("\n")
t.show_level_tree(t.root)

# Проверяем удаление узла с двумя потомками (например, 5)
t.del_node(5)

print("\nДерево после удаления узла 5:")
t.show_tree(t.root)
print("\n")
t.show_level_tree(t.root)

File: Data_Structures/binary_tree.cpp
#include <iostream>
#include <queue>
#include <utility> 


template <typename T>
class Node {
public:
    T data;
    Node<T> * left = nullptr;
    Node<T> * right = nullptr;
    
    Node(T data) : data(data) { }
    Node(T data, Node<T> * left, Node<T> * right) : data(data), left(left), right(right) { }
    
    ~Node() { }
};

template <typename T>
class Tree {
public:
    Node<T> * root = nullptr;
    Tree() = default;
    Tree(Node<T> * node) {
        root = node;
    }
    
    std::pair<Node<T>*, Node<T>*> find(Node<T>* child, Node<T>* parent, T value) {
        if (child == nullptr || child->data == value) {
            return {child, parent};
        } else if (child->data < value) {
            return find(child->right, child, value);
        }
        return find(child->left, child, value);
    }
    
    void add(Node<T>* node) {
        if (root == nullptr) {
            root = node;
            return;
        }
        auto [child, parent] = find(root, nullptr, node->data);
        if (child != nullptr) {
            return;
        }
        if (node->data < parent->data) {
            parent->left = node;
        } else {
            parent->right = node;
        }
        
    }
    
    void delete_leaf(Node<T>* child, Node<T>* parent) {
        if (parent == nullptr){
            root = nullptr;
        } else if (parent->left == child) {
            parent->left = nullptr;
        } else {
            parent->right = nullptr;
        }
        delete child;
        return;
    }
    
    void delete_one_child(Node<T>* child, Node<T>* parent) {
        Node<T>* nextNode = (child->left != nullptr) ? child->left : child->right;

        if (parent == nullptr) {
            root = nullptr;
        } else if (parent->left == child) {
            parent->left = nextNode;
        } else {
            parent->right = nextNode;
        }
        delete child;
        return;
    }

    Node<T>* find_min(Node<T>* node) {
        if (node->left) {
            return find_min(node->left);
        }
        return node;
    }
    
    void del_node(T key) {
        auto [child, parent] = find(root, nullptr, key);
        
        // not found
        if (child == nullptr) {
            return;
        }
        
        // leave node only
        if (child->left == nullptr && child->right == nullptr) {
            delete_leaf(child, parent);
            return;
        }
        // left child only
        if (child->right == nullptr || child->right == nullptr) {
            delete_one_child(child, parent);
            return ;
        }
        Node<T>* minNode = find_min(child->right);
        T minData = minNode->data;

        del_node(minData);
        child->data = minData;
    }
    
    void pprint(Node<T>* node) {
        if (node == nullptr) {
            return;
        }
        pprint(node->left);
        std::cout << node->data << " ";
        pprint(node->right);
    }
    
    void pprint_levels(Node<T>* node) {
        std::deque<Node<T>*> q;
        q.push_back(node);
        
        while (!q.empty()) {
            int lenLevel = q.size();
            for (int i = 0; i < lenLevel; ++i) {
                Node<T>* node = q.front();
                q.pop_front();
                std::cout << node->data << " ";
                if (node->left) {
                    q.push_back(node->left);
                }
                if (node->right) {
                    q.push_back(node->right);
                }
            }
            std::cout << std::endl;
        }
        
    }
    
};

int main() {
    Node<int>* left = new Node(3);
    Node<int>* right = new Node(6);
    Node<int>* node = new Node(5);
    Tree<int> t(node);
    t.add(left);
    t.add(right);
    t.add(new Node(4));
    t.add(new Node(2));
    
    t.pprint(t.root);
    t.del_node(3);
    std::cout << '\n' << std::endl;
    t.pprint_levels(t.root);
    
    return 0;
}
